zeroize/src/lib.rs

```diff
@@ -331,6 +333,27 @@ where
     }
 }
 
+/// Impl `Zeroize` on slices of MaybeUninit types
+/// This impl can eventually be optimized using an memset intrinsic,
+/// such as `core::intrinsics::volatile_set_memory`.
+/// This fills the slice with zeros
+/// Note that this ignore invariants that Z might have, becuase MaybeUninit removes all invariants.
+impl<Z> Zeroize for [MaybeUninit<Z>] {
+    fn zeroize(&mut self) {
+        let ptr = self.as_mut_ptr() as *mut MaybeUninit<u8>;
+        let size = self.len().checked_mul(mem::size_of::<Z>()).unwrap();
+        assert!(size <= core::isize::MAX as usize);
+        // Safety:
+        //
+        // This is safe, because every valid pointer is well aligned for u8
+        // and it is backed by a single allocated object for at least `self.len() * size_pf::<Z>()` bytes.
+        // and 0 is a valid value for `MaybeUninit<Z>`
+        // The memory of the slice should not wrap around the address space.
+        unsafe { volatile_set(ptr, MaybeUninit::new(0), size) }
```

Comment:
I'm still unclear on if #782 is UB and if it is, if this would also be UB

Reply:
This is `MaybeUninit` it has no invariants, even uninitialized values is a valid data for it.
It might be UB if a user say takes an enum and transmutes its unique reference to a `[MaybeUninit<Z>]` reference and then zeroize that, because the original enum still exists.
But by itself I do believe that this specific function is 100% UB safe.